### PR TITLE
[consensus] slow down if no user txns

### PR DESCRIPTION
**Pull request by zekun000**
## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/libra/developers.libra.org, and link to your PR here.)

### CODE REVIEW

File: consensus/src/txn_manager.rs [hunk 1]
```diff
@@ -50,28 +52,37 @@ impl TxnManager for MempoolProxy {
...
-            .map_err(anyhow::Error::from)?;
-        // wait for response
-        match monitor!(
-            "pull_txn",
-            timeout(Duration::from_secs(1), callback_rcv).await
-        ) {
-            Err(_) => {
-                Err(format_err!("[consensus] did not receive GetBlockResponse on time").into())
+        let no_pending_txns = exclude_txns.is_empty();
+        for _ in 0..50 {
```
Comment (davidiw) [inline consensus/src/txn_manager.rs:56]: a small doc block here would be priceless!

File: consensus/src/txn_manager.rs [hunk 2]
```diff
@@ -50,28 +52,37 @@ impl TxnManager for MempoolProxy {
...
+                ))),
+                Ok(resp) => match resp.map_err(anyhow::Error::from)?? {
+                    ConsensusResponse::GetBlockResponse(txns) => Ok(txns),
+                    _ => Err(MempoolError::from(anyhow::anyhow!(
+                        "[consensus] did not receive expected GetBlockResponse"
+                    ))),
+                },
+            }?;
+            if txns.is_empty() && no_pending_txns {
+                delay_for(Duration::from_millis(NO_TXN_DELAY)).await;
```
Comment (davidiw) [inline consensus/src/txn_manager.rs:80]: 500 ms delay, how will this translate in practice on testnet or a local run for example?
can we make this configurable? and if we do, could we make the configuration call into two different functions?
that way we can disable it if it becomes problematic on mainnet
also could a 500 ms delay ever translate into a timeout? should instead should we aim for a 1/10 or 1/4 time of the expected timeout (assuming that is relatively universal)
final thought... should we do 5 * 100 how much cpu are we burning doing this 50 times versus 5?

**Review by davidiw — COMMENTED** [review]
cool idea

**Comment by davidiw** [comment]
Someone said this is one line of code :P

**Comment by zekun000** [comment]
it's just for experiment (unless you really want it)

**Comment by davidiw** [comment]
> it's just for experiment (unless you really want it)

I think this is super useful for running a local testnet, our global testnet, and devnets as well... curious about @aching's thoughts.

**Comment by davidiw** [comment]
also curious about @metajack's thoughts as he's been working with partners.

**Comment by zekun000** [comment]
this seems acceptable now "fixed tps 10 : 10 TPS, 762 ms latency, 1000 ms p99 latency, no expired txns" with poll_count = 10, I'll default the value to 1 and delegate to each deployment to decide what value to use

File: consensus/src/txn_manager.rs [hunk 3]
```diff
@@ -50,28 +62,40 @@ impl TxnManager for MempoolProxy {
...
+                    "[consensus] did not receive GetBlockResponse on time"
+                ))),
+                Ok(resp) => match resp.map_err(anyhow::Error::from)?? {
+                    ConsensusResponse::GetBlockResponse(txns) => Ok(txns),
+                    _ => Err(MempoolError::from(anyhow::anyhow!(
+                        "[consensus] did not receive expected GetBlockResponse"
+                    ))),
+                },
+            }?;
+            if txns.is_empty() && no_pending_txns {
```
Comment (davidiw) [inline consensus/src/txn_manager.rs:90]: problem is you're gonna hit this on i == self.poll_count - 1, and you don't want to

**Review by davidiw — APPROVED** [review]
looks good but if we set this to 1, we ideally don't have an arbitrary sleep at the end ;)

File: consensus/src/txn_manager.rs (hunk 2 above)
Comment (aching) [reply consensus/src/txn_manager.rs:101]: Why didn't we make this configurable btw?

**Comment by metajack** [comment]
This looks excellent for local testing where you don't need high TPS and you'd rather your machine not sound like an airplane.

**Pull request closed without merging**